File: API/0425_Grid_Bot/GridBotStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using System.Collections.Generic;
using System.Linq;

using Ecng.Common;

using StockSharp.Algo;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

/// <summary>
/// Grid Bot Strategy
/// </summary>
public class GridBotStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleTypeParam;
	private readonly StrategyParam<decimal> _upperLimit;
	private readonly StrategyParam<decimal> _lowerLimit;
	private readonly StrategyParam<int> _gridCount;
	private readonly StrategyParam<int> _marketDirection;
	private readonly StrategyParam<bool> _useExtremes;

	private decimal _gridInterval;
	private decimal[] _gridLevels;
	private int _lastSignal;
	private int _lastSignalIndex;
	private decimal _signalLine;
	private ICandleMessage _previousCandle;

	public GridBotStrategy()
	{
		_candleTypeParam = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle type", "Candle type for strategy calculation.", "General");

		_upperLimit = Param(nameof(UpperLimit), 48000m)
			.SetDisplay("Upper Limit", "Grid upper boundary", "Grid Settings");

		_lowerLimit = Param(nameof(LowerLimit), 45000m)
			.SetDisplay("Lower Limit", "Grid lower boundary", "Grid Settings");

		_gridCount = Param(nameof(GridCount), 10)
			.SetDisplay("Grid Count", "Number of grid levels", "Grid Settings");

		_marketDirection = Param(nameof(MarketDirection), 0)
			.SetDisplay("Market Direction", "1=Up, 0=Neutral, -1=Down", "Strategy");

		_useExtremes = Param(nameof(UseExtremes), true)
			.SetDisplay("Use Extremes", "Use High/Low for signals", "Strategy");
	}

	public DataType CandleType
	{
		get => _candleTypeParam.Value;
		set => _candleTypeParam.Value = value;
	}

	public decimal UpperLimit
	{
		get => _upperLimit.Value;
		set => _upperLimit.Value = value;
	}

	public decimal LowerLimit
	{
		get => _lowerLimit.Value;
		set => _lowerLimit.Value = value;
	}

	public int GridCount
	{
		get => _gridCount.Value;
		set => _gridCount.Value = value;
	}

	public int MarketDirection
	{
		get => _marketDirection.Value;
		set => _marketDirection.Value = value;
	}

	public bool UseExtremes
	{
		get => _useExtremes.Value;
		set => _useExtremes.Value = value;
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> new[] { (Security, CandleType) };

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		// Calculate grid parameters
		var gridRange = UpperLimit - LowerLimit;
		_gridInterval = gridRange / GridCount;

		// Initialize grid levels
		_gridLevels = new decimal[GridCount + 1];
		for (var i = 0; i <= GridCount; i++)
		{
			_gridLevels[i] = LowerLimit + _gridInterval * i;
		}

		// Subscribe to candles
		var subscription = this.SubscribeCandles(CandleType);
		subscription
			.Bind(ProcessCandle)
			.Start();

		// Setup chart
		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		// Skip if strategy is not ready
		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		// Skip non-finished candles
		if (candle.State != CandleStates.Finished)
			return;

		var buyIndex = GetBuyLineIndex(candle);
		var sellIndex = GetSellLineIndex(candle);

		var buy = false;
		var sell = false;

		// Check for buy signal
		if (buyIndex > 0)
		{
			// No repeat trades at current level
			if (UseExtremes)
			{
				if (candle.LowPrice < _signalLine - _gridInterval)
					buy = true;
			}
			else
			{
				if (candle.ClosePrice < _signalLine - _gridInterval)
					buy = true;
			}

			// No trades outside of grid limits
			if (candle.ClosePrice >= UpperLimit || candle.ClosePrice < LowerLimit)
				buy = false;

			// Direction Filter (skip one signal if against market direction)
			if (MarketDirection == -1 && candle.LowPrice >= _signalLine - _gridInterval * 2)
				buy = false;
		}

		// Check for sell signal
		if (sellIndex > 0)
		{
			// No repeat trades at current level
			if (UseExtremes)
			{
				if (candle.HighPrice > _signalLine + _gridInterval)
					sell = true;
			}
			else
			{
				if (candle.ClosePrice > _signalLine + _gridInterval)
					sell = true;
			}

			// No trades outside of grid limits
			if (candle.ClosePrice <= LowerLimit || candle.ClosePrice > UpperLimit)
				sell = false;

			// Direction Filter (skip one signal if against market direction)
			if (MarketDirection == 1 && candle.HighPrice <= _signalLine + _gridInterval * 2)
				sell = false;
		}

		// Update trackers
		if (buy)
		{
			_lastSignal = 1;
			_lastSignalIndex = buyIndex;
			_signalLine = LowerLimit + _gridInterval * _lastSignalIndex;

			// Execute buy order
			if (Position <= 0)
			{
				var volume = Volume + Math.Abs(Position);
				BuyMarket(volume);
			}
		}
		else if (sell)
		{
			_lastSignal = -1;
			_lastSignalIndex = sellIndex;
			_signalLine = LowerLimit + _gridInterval * _lastSignalIndex;

			// Execute sell order
			if (Position >= 0)
			{
				var volume = Volume + Math.Abs(Position);
				SellMarket(volume);
			}
		}

		// Store previous candle for next iteration
		_previousCandle = candle;
	}

	private int GetBuyLineIndex(ICandleMessage candle)
	{
		var index = 0;

		for (var i = 0; i <= GridCount; i++)
		{
			var buyValue = _gridLevels[i];

			if (UseExtremes)
			{
				if (_previousCandle?.HighPrice > buyValue && candle.LowPrice <= buyValue)
					index = i;
			}
			else
			{
				if (_previousCandle?.ClosePrice > buyValue && candle.ClosePrice <= buyValue)
					index = i;
			}
		}

		return index;
	}

	private int GetSellLineIndex(ICandleMessage candle)
	{
		var index = 0;

		for (var i = 0; i <= GridCount; i++)
		{
			var sellValue = _gridLevels[i];

			if (UseExtremes)
			{
				if (_previousCandle?.LowPrice < sellValue && candle.HighPrice >= sellValue)
					index = i;
			}
			else
			{
				if (_previousCandle?.ClosePrice < sellValue && candle.ClosePrice >= sellValue)
					index = i;
			}
		}

		return index;
	}
}

File: API/0425_Grid_Bot/PY/grid_bot_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage, AverageTrueRange
from StockSharp.Algo.Strategies import Strategy


class grid_bot_strategy(Strategy):
    """Grid Bot Strategy. Dynamic grid around EMA with ATR spacing."""

    def __init__(self):
        super(grid_bot_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(30))) \
            .SetDisplay("Candle type", "Candle type for strategy calculation.", "General")
        self._ma_length = self.Param("MALength", 50) \
            .SetDisplay("MA Length", "Moving average for grid center", "Grid Settings")
        self._atr_length = self.Param("ATRLength", 14) \
            .SetDisplay("ATR Length", "ATR period for grid spacing", "Grid Settings")
        self._grid_count = self.Param("GridCount", 3) \
            .SetDisplay("Grid Count", "Number of grid levels each side", "Grid Settings")
        self._grid_multiplier = self.Param("GridMultiplier", 0.5) \
            .SetDisplay("Grid Multiplier", "ATR multiplier for grid spacing", "Grid Settings")
        self._cooldown_bars = self.Param("CooldownBars", 20) \
            .SetDisplay("Cooldown Bars", "Bars to wait between trades", "Risk")

        self._ma = None
        self._atr = None
        self._prev_close = 0.0
        self._cooldown_remaining = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(grid_bot_strategy, self).OnReseted()
        self._ma = None
        self._atr = None
        self._prev_close = 0.0
        self._cooldown_remaining = 0

    def OnStarted(self, time):
        super(grid_bot_strategy, self).OnStarted(time)

        self._ma = ExponentialMovingAverage()
        self._ma.Length = int(self._ma_length.Value)

        self._atr = AverageTrueRange()
        self._atr.Length = int(self._atr_length.Value)

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._ma, self._atr, self._on_process).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, self._ma)
            self.DrawOwnTrades(area)

    def _on_process(self, candle, ma_val, atr_val):
        if candle.State != CandleStates.Finished:
            return

        if not self._ma.IsFormed or not self._atr.IsFormed:
            self._prev_close = float(candle.ClosePrice)
            return

        if not self.IsFormedAndOnlineAndAllowTrading():
            self._prev_close = float(candle.ClosePrice)
            return

        if self._cooldown_remaining > 0:
            self._cooldown_remaining -= 1
            self._prev_close = float(candle.ClosePrice)
            return

        if self._prev_close == 0.0 or float(atr_val) <= 0:
            self._prev_close = float(candle.ClosePrice)
            return

        close = float(candle.ClosePrice)
        ma = float(ma_val)
        atr = float(atr_val)
        grid_mult = float(self._grid_multiplier.Value)
        grid_count = int(self._grid_count.Value)
        cooldown = int(self._cooldown_bars.Value)
        grid_spacing = atr * grid_mult

        for i in range(1, grid_count + 1):
            lower_grid = ma - grid_spacing * i
            upper_grid = ma + grid_spacing * i

            if self._prev_close > lower_grid and close <= lower_grid and self.Position <= 0:
                if self.Position < 0:
                    self.BuyMarket(Math.Abs(self.Position))
                self.BuyMarket(self.Volume)
                self._cooldown_remaining = cooldown
                self._prev_close = close
                return

            if self._prev_close < upper_grid and close >= upper_grid and self.Position >= 0:
                if self.Position > 0:
                    self.SellMarket(Math.Abs(self.Position))
                self.SellMarket(self.Volume)
                self._cooldown_remaining = cooldown
                self._prev_close = close
                return

        # Mean reversion exits at MA
        if self.Position > 0 and self._prev_close < ma and close >= ma:
            self.SellMarket(Math.Abs(self.Position))
            self._cooldown_remaining = cooldown
        elif self.Position < 0 and self._prev_close > ma and close <= ma:
            self.BuyMarket(Math.Abs(self.Position))
            self._cooldown_remaining = cooldown

        self._prev_close = close

    def CreateClone(self):
        return grid_bot_strategy()
